File: shared-libs/ts-commons/src/money/money.ts
export interface Money {
  readonly minorUnits: number;
  readonly currency: string;
}

const isoCurrency = /^[A-Z]{3}$/;

export function money(minorUnits: number, currency: string): Money {
  if (!Number.isSafeInteger(minorUnits)) {
    throw new Error("minorUnits must be a safe integer");
  }
  const c = currency?.toUpperCase();
  if (!c || !isoCurrency.test(c)) {
    throw new Error("currency must be ISO-4217 3-letter code");
  }
  return Object.freeze({ minorUnits, currency: c });
}

export const usd = (cents: number): Money => money(cents, "USD");

export function plus(a: Money, b: Money): Money {
  if (a.currency !== b.currency) throw new Error(`currency mismatch ${a.currency}/${b.currency}`);
  return money(a.minorUnits + b.minorUnits, a.currency);
}

export function minus(a: Money, b: Money): Money {
  if (a.currency !== b.currency) throw new Error(`currency mismatch ${a.currency}/${b.currency}`);
  return money(a.minorUnits - b.minorUnits, a.currency);
}

export const isNegative = (m: Money): boolean => m.minorUnits < 0;
export const isPositive = (m: Money): boolean => m.minorUnits > 0;
export const isZero     = (m: Money): boolean => m.minorUnits === 0;

export function formatMoney(m: Money): string {
  const sign = m.minorUnits < 0 ? "-" : "";
  const abs = Math.abs(m.minorUnits);
  const major = Math.floor(abs / 100);
  const minor = (abs % 100).toString().padStart(2, "0");
  return `${m.currency} ${sign}${major}.${minor}`;
}

File: shared-libs/java-commons/src/main/java/com/finspoly/commons/money/Money.java
package com.finspoly.commons.money;

import java.util.Objects;

/**
 * Immutable monetary amount stored as minor units (cents). Never use float/double for money.
 */
public final class Money {
    private final long minorUnits;
    private final String currency;

    public Money(long minorUnits, String currency) {
        if (currency == null || currency.length() != 3) {
            throw new IllegalArgumentException("currency must be ISO-4217 3-letter code");
        }
        this.minorUnits = minorUnits;
        this.currency = currency.toUpperCase();
    }

    public static Money usd(long cents) {
        return new Money(cents, "USD");
    }

    public static Money zero(String currency) {
        return new Money(0, currency);
    }

    public long minorUnits() { return minorUnits; }
    public String currency() { return currency; }

    public Money plus(Money other) {
        requireSameCurrency(other);
        return new Money(Math.addExact(minorUnits, other.minorUnits), currency);
    }

    public Money minus(Money other) {
        requireSameCurrency(other);
        return new Money(Math.subtractExact(minorUnits, other.minorUnits), currency);
    }

    public Money negate() {
        return new Money(-minorUnits, currency);
    }

    public boolean isNegative() { return minorUnits < 0; }
    public boolean isPositive() { return minorUnits > 0; }
    public boolean isZero() { return minorUnits == 0; }

    public int compareTo(Money other) {
        requireSameCurrency(other);
        return Long.compare(minorUnits, other.minorUnits);
    }

    private void requireSameCurrency(Money other) {
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException("currency mismatch: " + currency + " vs " + other.currency);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Money)) return false;
        Money m = (Money) o;
        return minorUnits == m.minorUnits && currency.equals(m.currency);
    }

    @Override
    public int hashCode() { return Objects.hash(minorUnits, currency); }

    @Override
    public String toString() {
        return String.format("%s %d.%02d", currency, minorUnits / 100, Math.abs(minorUnits % 100));
    }
}
